src/celeritas/io/RootStepWriter.cc

```diff
@@ -0,0 +1,247 @@
+/*!
+ * Create steps tree. In order to have the option to individually select any
+ * member of `StepStateData` (StepData.hh) to be stored into the ROOT file, we
+ * cannot store an MC truth step object. This is accomplished by "flattening"
+ * the data so that each member of `TStepData` (MCTruthData.hh) is an
+ * individual branch that can be created based on the `StepSelection` booleans.
+ *
+ * To simplify the process of moving from Collection to a ROOT branch
+ * `TStepData` members *must* have the *exact* same name as the Collection
+ * members of `StepStateData`. This allows the use of C++ macros to
+ * automatically create ROOT branches with the same names as the struct
+ * elements.
+ */
+void RootStepWriter::make_tree()
+{
+#define CREATE_BRANCH_IF_SELECTED(ATTR)                     \
+    do                                                      \
+    {                                                       \
+        if (this->selection_.ATTR)                          \
+        {                                                   \
+            this->tstep_tree_->Branch(#ATTR, &tstep_.ATTR); \
+        }                                                   \
+    } while (0)
+
+#define CREATE_POINT_BRANCH_IF_SELECTED(ATTR, NAME, REF, TYPE)           \
+    do                                                                   \
+    {                                                                    \
+        if (this->selection_.ATTR)                                       \
+        {                                                                \
+            std::string leaf_def(NAME);                                  \
+            std::string str_attr(#ATTR);                                 \
+            const auto  last_p = str_attr.find_last_of(".");             \
+            if (str_attr.find("pos", last_p, 3) != std::string::npos     \
+                || str_attr.find("dir", last_p, 3) != std::string::npos) \
+            {                                                            \
+                leaf_def += "[3]";                                       \
+            }                                                            \
+            leaf_def += "/" + std::string(TYPE);                         \
+            this->tstep_tree_->Branch(NAME, &REF, leaf_def.c_str());     \
+        }                                                                \
+    } while (0)
+
+    tstep_tree_.reset(new TTree("steps", "steps"));
+    tstep_tree_->Branch("track_id", &tstep_.track_id); // Always on
+
+    // Step selection data
+    CREATE_BRANCH_IF_SELECTED(event_id);
+    CREATE_BRANCH_IF_SELECTED(track_step_count);
+    CREATE_BRANCH_IF_SELECTED(action_id);
+    CREATE_BRANCH_IF_SELECTED(step_length);
+    CREATE_BRANCH_IF_SELECTED(particle);
+    CREATE_BRANCH_IF_SELECTED(energy_deposition);
+
+    // Step point selection data //
+    // Pre-step
+    {
```

Comment:
Seems like you could loop over this pretty easily by concatenating some strings.